core/src/main/java/com/linecorp/armeria/internal/common/PathAndQuery.java

```diff
@@ -142,6 +143,8 @@ public void storeInCache(@Nullable String rawPath) {
     @Nullable
     private final String query;
 
+    private volatile boolean cached;
```

Comment:
Probably no need to use `volatile` because what we need is just an approximation here, i.e. it will still dramatically reduce the chance of double-putting.

Reply:
Agreed. `cached` does not need to tightly guarantee single-putting.